### PR TITLE
[cs_pinyin] CS-Pinyin keyboard initial public version

### PR DESCRIPTION
**Pull request by mcdurdin**
This is a public version of the CS-Pinyin keyboard (from keyman-keyboards-internal). It contains all the files necessary to build the keyboard and the DLLs that go with it.

~~I will be checking the build agent to ensure that it can build the DLLs.~~ We have checked in builds of the DLLs so that we don't introduce a Windows dependency for building the keyboards repo.

### CODE REVIEW

**Comment by mcdurdin** [comment]
* .cfx file has the wrong name (should match keyboard base name).
* Even when .cfx is named correctly, the keyboard isn't working.

**Comment by LornaSIL** [comment]
When I try to build I get this error:

```
cs_pinyin.kps: Warning: Warning: File ...\keyboards\cs_pinyin\source\..\build\KeymnIMX.dll does not exist.
cs_pinyin.kps: Warning: Warning: File ...\keyboards\cs_pinyin\source\..\build\imxconfig.exe does not exist.
cs_pinyin.kps: Warning: Warning: File ...\keyboards\cs_pinyin\source\..\build\KeymnIMX.x64.dll does not exist.
```
I think the .kps is looking in the wrong place for the .dlls?

A lot of the documentation still talks about Tavultesoft and Keyman 6.

**Comment by mcdurdin** [comment]
I've just added the work-in-progress label to clarify that this keyboard is not yet really ready for review. The content was put online to give another developer access to the source for their own purposes and I will continue to prepare this for release as time permits.

Do you think we should close the PR and reopen it later? Or just leave it open as a hint to yours truly that something should be done about it?

**Comment by LornaSIL** [comment]
The work in progress label is helpful.

**Comment by MakaraSok** [comment]
Missing files in the package have been copied from the legacy .kmp and put in the /build folder. Those files are: imxconfig.exe, KeymnIMX.dll and KeymnIMX.x64.dll.

The tweaks fix the errors, but the keyboard doesn't function well yet. It needs further investigation.

Also, the content of the README.md and CS-Pinyin ReadMe.html have been updated.

**Comment by mcdurdin** [comment]
Note: those files are supposed to be built during the keyboard build
process but the build script needs updating to make it work with the new
build.sh infrastructure.

**Comment by DavidLRowe** [comment]
@mcdurdin Two years ago you wrote: "Do you think we should close the PR and reopen it later? Or just leave it open as a hint to yours truly that something should be done about it?" How well has that hint been working? ;)

**Comment by DavidLRowe** [comment]
Related: #219

**Comment by mcdurdin** [comment]
@DavidLRowe ask me again in two years. Currently my to do list is probably about 5 years long and getting longer.

File: release/c/cs_pinyin/source/docs/CS-Pinyin ReadMe.html [hunk 1]
```diff
@@ -0,0 +1,56 @@
+<html><head><title>Simplified Chinese Input Method</title>
+<style>
+body {
+	margin: 30px; 
+	font-family: 'Arial Unicode MS', Helvetica, sans-serif; 
+	font-size: 10pt;color: #2D2C2C;}
+img {border: none; }
```
Comment (darcywong00) [inline release/c/cs_pinyin/source/docs/CS-Pinyin ReadMe.html:7]: FYI if you resume this PR, I removed this styling block on keymanapp/help.keyman.com#923 and it seemed to fix  the page layout

**Comment by rc-swag** [comment]
Since we need commit [d5a878](https://github.com/keymanapp/keyboards/pull/64/commits/258b32b394dd09e66510df0fea644c3371d5a878) to guarnteed the simplified Chinese keyboard works in more then one open application. This PR needs the work done to get it accepted and merged. 
TODO:

- [ ] introduce build.sh so the KeymnIMX.dll, KeymnIMX.x64.dll and imxconfig.exe projects can be built on the build agents
- [ ] Update documentation references to specific Keyman Versions, and Tavultesoft where required. See previous comments in this PR thread.

**Comment by LornaSIL** [comment]
> ask me again in two years. Currently my to do list is probably about 5 years long and getting longer.

Four years ago on June 18, 2020 you asked us to ask you in 2 years. Is there any likelihood this will be done @mcdurdin?

**Comment by rc-swag** [comment]
> Four years ago on June 18, 2020 you asked us to ask you in 2 years. Is there any likelihood this will be done @mcdurdin?

Well, I have fixed a problem with it when you open more than one application at a time. However, the main issue the first point of my TODO above is the main sticking point it needs to build the dll. I think it will be done as we get the single repo keyboards?

**Comment by mcdurdin** [comment]
@rc-swag and I will be discussing cs-pinyin this week with the aim of getting something published, even if it is not optimal. (A full IME/picker rework is on the future roadmap)

**Comment by mcdurdin** [comment]
> Is there any likelihood this will be done @mcdurdin?

It's nearly done now... we should be able to get it over the line soon I hope!

**Comment by DavidLRowe** [comment]
I get it, @mcdurdin : You're holding out for a few more weeks to make it a seven-year PR!

**Comment by LornaSIL** [comment]
- Could you delete `keyboard_info`?
- In `LICENSE.md` please extend the copyright date 2017-2024 - actually it should be `2003-2024`.
- In `HISTORY.md` are you interested in bumping the date to 2024 and our goal could be a release on July 27 :)
- `README.md` could have copyright and version removed.
- Could you change the encoding for .kmn from ANSI to UTF-8?
   - Please remove copyright year in .kmn
- In the .kps, **Details** you could remove the version and tick "Package version follows keyboard version"
   - Please remove copyright date or extend it to 2024.
- CS-Pinyin Install.bmp needs updating to remove the "(c) 2011 Tavultesoft"
- welcome.htm could have the copyright information removed
- `CS-Pinyin ReadMe.html` it would be wise to rename this to `readme.htm` since it has spaces and casing. Then it needs updating in the .kps.
   - This file could also have the copyright information removed
- Since this is exactly the same name as the keyboard in `legacy`, there should not be a deprecation, the folder in `legacy` should be deleted.

**Comment by LornaSIL** [comment]
> Since this is exactly the same name as the keyboard in legacy, there should not be a deprecation, the folder in legacy should be deleted.

Scratch that. It has a different folder name. So, the .kps file needs to be updated to deprecate the `legacy\c\cs-pinyin` keyboard.

**Comment by LornaSIL** [comment]
I've done minor testing of the keyboard in Word 2019. It seems to be working, even in Github comments (特命). I'm not sure how else to test.

I wondered if you were wanting to address this issue? https://github.com/keymanapp/keyboards/issues/219

Also, all the documentation (pdfs) are outdated, but I suspect that is out of scope for this current fix.

**Comment by mcdurdin** [comment]
> I wondered if you were wanting to address this issue? #219

This should definitely be fixed in a separate PR!

> Also, all the documentation (pdfs) are outdated, but I suspect that is out of scope for this current fix.

Yes, I think we'll aim to update documentation separately as well.

For remaining items, @rc-swag are you able to cover those? (What else are we waiting on for publishing this?)

**Comment by rc-swag** [comment]
> For remaining items, @rc-swag are you able to cover those? (What else are we waiting on for publishing this?)

Yes I can address those items @LornaSIL raised. The last two commits address the functional issues with the window ids, multi-monitor support and finally the registry settings.

**Comment by rc-swag** [comment]
- [ ] - Could you delete `keyboard_info`?
- [ ] - In `LICENSE.md` please extend the copyright date 2017-2024 - actually it should be `2003-2024`.
- [ ] - In `HISTORY.md` are you interested in bumping the date to 2024 and our goal could be a release on July 27 :)
- [ ] - `README.md` could have copyright and version removed.
- [ ] - Could you change the encoding for .kmn from ANSI to UTF-8?
- [ ]    - Please remove copyright year in .kmn
- [ ] - In the .kps, **Details** you could remove the version and tick "Package version follows keyboard version"
- [ ]    - Please remove copyright date or extend it to 2024.
- [ ] - CS-Pinyin Install.bmp needs updating to remove the "(c) 2011 Tavultesoft"
- [ ] - welcome.htm could have the copyright information removed
- [ ] - `CS-Pinyin ReadMe.html` it would be wise to rename this to `readme.htm` since it has spaces and casing. Then it needs updating in the .kps.
- [ ]    - This file could also have the copyright information removed

**Comment by rc-swag** [comment]
@LornaSIL  and @mcdurdin  I believe I have addressed those remaining items.

**Comment by LornaSIL** [comment]
Just one more thing:

The .kps file needs to be updated to deprecate the `legacy\c\cs-pinyin` keyboard.

That is in **Details** / **Related Packages**

Click on **Add** and type in `cs-pinyin` for the Related package ID. Select **Deprecated**. Apparently it finds the path by itself.

image

**Comment by rc-swag** [comment]
Ok added the related package

**Review by LornaSIL — APPROVED** [review]
LGTM!!!!

**Pull request merged**